plugins/modules/opkg.py

```diff
@@ -97,144 +97,108 @@
-        else:
-            version_str = ""
-
-        if force:
-            rc, out, err = module.run_command([opkg_path, "install", force, package + version_str])
-        else:
-            rc, out, err = module.run_command([opkg_path, "install", package + version_str])
-
-        if not query_package(module, opkg_path, package, version):
-            module.fail_json(msg="failed to install %s%s: %s" % (package, version_str, out))
-
-        install_c += 1
-
-    if install_c > 0:
-        module.exit_json(changed=True, msg="installed %s package(s)" % (install_c))
-
-    module.exit_json(changed=False, msg="package(s) already present")
-
-
-def main():
-    module = AnsibleModule(
+class Opkg(StateModuleHelper):
+    module = dict(
         argument_spec=dict(
             name=dict(aliases=["pkg"], required=True, type="list", elements="str"),
             state=dict(default="present", choices=["present", "installed", "absent", "removed"]),
             force=dict(default="", choices=["", "depends", "maintainer", "reinstall", "overwrite", "downgrade", "space", "postinstall", "remove",
                                             "checksum", "removal-of-dependent-packages"]),
             update_cache=dict(default=False, type='bool'),
-        )
+        ),
     )
 
-    opkg_path = module.get_bin_path('opkg', True, ['/bin'])
+    def __init_module__(self):
+        self.vars.set("install_c", 0, output=False, change=True)
+        self.vars.set("remove_c", 0, output=False, change=True)
 
-    p = module.params
+        state_map = dict(
+            query="list-installed",
+            present="install",
+            installed="install",
+            absent="remove",
+            removed="remove",
+        )
 
-    if p["update_cache"]:
-        update_package_db(module, opkg_path)
+        def _force(value):
+            if value == "":
+                value = None
+            return cmd_runner_fmt.as_optval("--force-")(value)
+
+        self.runner = CmdRunner(
+            self.module,
+            command="opkg",
+            arg_formats=dict(
+                package=cmd_runner_fmt.as_list(),
+                state=cmd_runner_fmt.as_map(state_map),
+                force=cmd_runner_fmt.as_func(_force),
+                update_cache=cmd_runner_fmt.as_bool("update")
+            ),
+        )
 
-    pkgs = p["name"]
+    @staticmethod
+    def split_name_and_version(package):
+        """ Split the name and the version when using the NAME=VERSION syntax """
+        splitted = package.split('=', 1)
+        if len(splitted) == 1:
+            return splitted[0], None
+        else:
+            return splitted[0], splitted[1]
+
+    def _package_in_desired_state(self, name, desired_installed, version=None):
+        dummy, out, dummy = self.runner("state package").run(state="query", package=name)
+        if version is None:
+            version = ""
+
+        has_package = out.startswith(name + " - %s" % version)
```

Comment:
`name + " - %s"` will catch not only 
`foo - 1.0` but also `foo - 1.0f1`.
It should be `name + " - %s "` if `version is not None`